import "BuildInfoPB.proto";

package molecular_dynamics.glycoprotein_builder;
option java_package = "org.glycam.molecular_dynamics.glycoprotein_builder";

// This message is sent from C++ to Java,
message PdbInfo {
    repeated PdbResidueInfo his_residue = 1;
    repeated CYSPair close_cys_pair = 2;
    repeated GlycosylationSpot glycosylation_spot = 3;
}

// This is sent from Java to C++.
message PdbModificationInfo {
    required string pdb_file = 1;

    repeated PdbMapping his_mapping = 2;

    // This is separate from other mapping because we may need to insert bonds.
    repeated CYSPair close_cys_pair = 3;

    repeated GlycosylationInfo glycosylation = 4;
}

message PdbResidueInfo {
    // This is the index of the residue in the gmml::Structure. It is redundant
    // with the other 3, but both pieces of information may be useful. However,
    // the redundancy should be reevaluated at some point.
    required int32 index = 1;

    required string chain_id = 2;
    required int32 res_num = 3;
    required string i_code = 4;
}

message PdbMapping {
    required PdbResidueInfo residue = 1;
    required string mapped_name = 2;
}

message CYSPair {
    required PdbResidueInfo cys1 = 1;
    required PdbResidueInfo cys2 = 2;
    required double distance = 3;
    required bool bonded = 4;
}

message GlycosylationSpot {
    required PdbResidueInfo info = 1;
    required string name = 2;
}

message GlycosylationInfo {
    required PdbResidueInfo spot = 1;
    required molecular_dynamics.oligosaccharide_builder.BuildInfo glycan = 2;
}
